code/modules/projectiles/tgprojectile.dm

```diff
@@ -0,0 +1,1276 @@
+
+/obj/item/projectile/Crossed(atom/movable/AM) //A mob moving on a tile with a projectile is hit by it.
+	. = ..()
+	if(isliving(AM) && !(flags_pass & PASSMOB))
+		var/mob/living/L = AM
+		if(can_hit_target(L, permutated, (AM == original)))
+			Bump(AM)
+
+/obj/item/projectile/Move(atom/newloc, dir = NONE)
+	. = ..()
+	if(!.)
+		return
+	if(!fired)
+		return
+	distance_travelled++
+	if(can_hit_target(original, permutated, TRUE))
+		Bump(original)
+	for(var/atom/i in (get_turf(newloc)).contents)
+		if(i.density && i.loc == loc)
+			Bump(i)
+
+/obj/item/projectile/Destroy()
+	if(hitscan)
+		finalize_hitscan_and_generate_tracers()
+	STOP_PROCESSING(SSprojectiles, src)
+	cleanup_beam_segments()
+	qdel(trajectory)
+	return ..()
+
+/obj/item/projectile/proc/cleanup_beam_segments()
+	QDEL_LIST_ASSOC(beam_segments)
+	beam_segments = list()
+	qdel(beam_index)
+
+/obj/item/projectile/proc/finalize_hitscan_and_generate_tracers(impacting = TRUE)
+	if(trajectory && beam_index)
+		var/datum/point/pcache = trajectory.copy_to()
+		beam_segments[beam_index] = pcache
+	generate_hitscan_tracers(null, null, impacting)
+
+/obj/item/projectile/proc/generate_hitscan_tracers(cleanup = TRUE, duration = 3, impacting = TRUE)
+	if(!length(beam_segments))
+		return
+	if(muzzle_type && duration > 0)
+		var/datum/point/p = beam_segments[1]
+		var/atom/movable/thing = new muzzle_type
+		p.move_atom_to_src(thing)
+		var/matrix/M = new
+		M.Turn(original_angle)
+		thing.transform = M
+		thing.color = color
+		//thing.set_light(muzzle_flash_range, muzzle_flash_intensity, muzzle_flash_color_override? muzzle_flash_color_override : color)
+		QDEL_IN(thing, duration)
+	if(impacting && impact_type && duration > 0)
+		var/datum/point/p = beam_segments[beam_segments[beam_segments.len]]
+		var/atom/movable/thing = new impact_type
+		p.move_atom_to_src(thing)
+		var/matrix/M = new
+		M.Turn(Angle)
+		thing.transform = M
+		thing.color = color
+		//thing.set_light(impact_light_range, impact_light_intensity, impact_light_color_override? impact_light_color_override : color)
+		QDEL_IN(thing, duration)
+	if(cleanup)
+		cleanup_beam_segments()
+
+// OLD PROCS, TRY TO REPLACE
+
+/mob/proc/bullet_message(obj/item/projectile/P)
+	if(!P) return
+
+	if(P.ammo.flags_ammo_behavior & AMMO_IS_SILENCED)
+		to_chat(src, "[isxeno(src) ? "<span class='xenodanger'>" : "<span class='highdanger'>" ]You've been shot in the [parse_zone(P.def_zone)] by [P.name]!</span>")
+	else
+		visible_message("<span class='danger'>[name] is hit by the [P.name] in the [parse_zone(P.def_zone)]!</span>", \
+						"<span class='highdanger'>You are hit by the [P.name] in the [parse_zone(P.def_zone)]!</span>", null, 4)
+
+	if(ismob(P.firer))
+		var/mob/firingMob = P.firer
+		var/turf/T = get_turf(firingMob)
+		if(ishuman(firingMob) && ishuman(src) && firingMob.mind && mind) //One human shot another, be worried about it but do everything basically the same //special_role should be null or an empty string if done correctly
```

Comment:
You may want to grab the lasest version of this from the master, I've fixed the way FF is done.